### PR TITLE
Show chat panel when opening a video room with unread messages

### PR DESCRIPTION
**Pull request by robintown**
This required some refactoring of RightPanelStore to get it working with both the new functionality and the tests.

Closes https://github.com/vector-im/element-web/issues/22527

---
Here's what your changelog entry will look like:

## ✨ Features
 * Show chat panel when opening a video room with unread messages ([\#8812](https://github.com/matrix-org/matrix-react-sdk/pull/8812)). Fixes vector-im/element-web#22527.

### CODE REVIEW

File: src/components/structures/RightPanel.tsx [hunk 1]
```diff
@@ -91,12 +91,6 @@ export default class RightPanel extends React.Component<IProps, IState> {
             currentCard = RightPanelStore.instance.currentCardForRoom(props.room.roomId);
         }
 
-        if (currentCard?.phase && !RightPanelStore.instance.isPhaseValid(currentCard.phase, !!props.room)) {
-            // XXX: We can probably get rid of this workaround once GroupView is dead, it's unmounting happens weirdly
-            // late causing the app to soft-crash due to lack of a room object being passed to a RightPanel
-            return null; // skip this update, we're about to be unmounted and don't have the appropriate props
-        }
```
Comment (t3chguy) [inline src/components/structures/RightPanel.tsx:98]: 🥳

**Review by t3chguy — APPROVED** [review]
Looks sane to me :shipit:

**Review by turt2live — CHANGES_REQUESTED** [review]
overall I'm quite terrified of this area - would love to see tests which prove most/all existing functionality isn't broken given we have none at the moment.

File: src/stores/right-panel/RightPanelStore.ts [hunk 2]
```diff
@@ -51,20 +51,21 @@ export default class RightPanelStore extends ReadyWatchingStore {
...
     private viewedRoomId: Optional<string>;
 
-    private constructor() {
+    // Public for tests
+    public constructor() {
         super(defaultDispatcher);
     }
 
-    protected async onReady(): Promise<any> {
+    protected onReady = async (): Promise<any> => {
```
Comment (turt2live) [inline src/stores/right-panel/RightPanelStore.ts:59]: why do these need to be bound functions?

File: test/stores/right-panel/RightPanelStore-test.ts [hunk 3]
```diff
@@ -0,0 +1,204 @@
...
+import { stubClient } from "../../test-utils";
+import { MatrixClientPeg } from "../../../src/MatrixClientPeg";
+import { Action } from "../../../src/dispatcher/actions";
+import defaultDispatcher from "../../../src/dispatcher/dispatcher";
+import { ActiveRoomChangedPayload } from "../../../src/dispatcher/payloads/ActiveRoomChangedPayload";
+import RightPanelStore from "../../../src/stores/right-panel/RightPanelStore";
+import { RightPanelPhases } from "../../../src/stores/right-panel/RightPanelStorePhases";
+import SettingsStore from "../../../src/settings/SettingsStore";
+
+describe("RightPanelStore", () => {
```
Comment (turt2live) [inline test/stores/right-panel/RightPanelStore-test.ts:29]: these tests are great and much-needed, though I was hoping to see tests which cover the phases as that's a much more fragile area. Specifically that the right panel shows the correct thing given different application states (defaults to info card, shows member list, doesn't show user info after switching rooms, etc)

File: src/stores/right-panel/RightPanelStore.ts (hunk 2 above)
Comment (robintown) [reply src/stores/right-panel/RightPanelStore.ts:59]: Because otherwise `onReady` gets called from *the constructor* of the parent class, which apparently interacts weirdly with the overriding and binding of `onReady`, causing a type error that `this.onVerificationRequestUpdate` is undefined.

File: test/stores/right-panel/RightPanelStore-test.ts (hunk 3 above)
Comment (robintown) [reply test/stores/right-panel/RightPanelStore-test.ts:31]: Tests added, with the exception of "defaults to info card", since that's not the actual current behavior of the right panel, nor was it before this PR.

File: src/stores/right-panel/RightPanelStore.ts (hunk 2 above)
Comment (turt2live) [reply src/stores/right-panel/RightPanelStore.ts:58]: err, that's extremely surprising. It being called from the parent class shouldn't cause `this` to change scope in a weird way like this. 

it also wasn't a bound function before, which has me quite curious if it ever worked. 

Do you have a stacktrace from when it failed?

**Review by turt2live — COMMENTED** [review]
otherwise lgtm - just the comment about bound functions from the last review which needs a bit of digging into.

File: src/stores/right-panel/RightPanelStore.ts (hunk 2 above)
Comment (robintown) [reply src/stores/right-panel/RightPanelStore.ts:58]: Yeah, it's confused me too, but if it's not bound, then 10 out of the 19 RightPanelStore tests fail in the `beforeEach` stage with:

```
TypeError [ERR_INVALID_ARG_TYPE]: The "listener" argument must be of type function. Received undefined

  58 |     protected async onReady(): Promise<any> {
  59 |         this.viewedRoomId = RoomViewStore.instance.getRoomId();
> 60 |         this.matrixClient.on(CryptoEvent.VerificationRequest, this.onVerificationRequestUpdate);
     |                           ^
  61 |         this.loadCacheFromSettings();
  62 |         this.emitAndUpdateSettings();
  63 |     };

  at RightPanelStore.onReady (src/stores/right-panel/RightPanelStore.ts:60:27)
  at new ReadyWatchingStore (src/stores/ReadyWatchingStore.ts:40:18)
  at new RightPanelStore (src/stores/right-panel/RightPanelStore.ts:55:9)
  at Object.<anonymous> (test/stores/right-panel/RightPanelStore-test.ts:45:17)
```

and I have no idea what to make of that. Same thing for the RightPanel test suite, in which both tests fail `at spinUpStores (test/components/structures/RightPanel-test.tsx:76:25)`.

Comment (robintown) [reply src/stores/right-panel/RightPanelStore.ts:58]: It hurts my brain to think about how that could possibly happen, but binding the methods seems to work and be benign, so :woman_shrugging:

Comment (turt2live) [reply src/stores/right-panel/RightPanelStore.ts:58]: The bound function seems to change the call path a bit. Walking through it with a debugger shows that the construction of a `RightPanelStore` sends it through the `super`'s `onReady`, which is the wrong `onReady`. It's only the manual call in the test which hits the actual `onReady` function defined by `RightPanelStore`. Switching it back to an unbound function means the `RightPanelStore`'s override is in fact called, though with a weird `this` reference which appears to be a different `RightPanelStore` entirely... 

/me pokes it some more.

Comment (turt2live) [reply src/stores/right-panel/RightPanelStore.ts:58]: Right, so it's because of how much of a nightmare our store layout is. 

TLDR: This fixes it:

```diff
diff --git a/src/stores/right-panel/RightPanelStore.ts b/src/stores/right-panel/RightPanelStore.ts
index 2cfafb26ac2..b3e46e7ff71 100644
--- a/src/stores/right-panel/RightPanelStore.ts
+++ b/src/stores/right-panel/RightPanelStore.ts
@@ -53,18 +53,28 @@ export default class RightPanelStore extends ReadyWatchingStore {

     private constructor() {
         super(defaultDispatcher);
+        this.reset();
     }

-    protected onReady = async (): Promise<any> => {
+    protected async onReady(): Promise<any> {
         this.viewedRoomId = RoomViewStore.instance.getRoomId();
         this.matrixClient.on(CryptoEvent.VerificationRequest, this.onVerificationRequestUpdate);
         this.loadCacheFromSettings();
         this.emitAndUpdateSettings();
-    };
+    }

-    protected onNotReady = async (): Promise<any> => {
+    protected async onNotReady(): Promise<any> {
         this.matrixClient.off(CryptoEvent.VerificationRequest, this.onVerificationRequestUpdate);
-    };
+    }
+
+    /**
+     * Resets the store. Intended for test usage only.
+     */
+    public reset() {
+        this.global = null;
+        this.byRoom = {};
+        this.viewedRoomId = null;
+    }

     protected onDispatcherAction(payload: ActionPayload) {
         if (payload.action !== Action.ActiveRoomChanged) return;
diff --git a/test/components/structures/RightPanel-test.tsx b/test/components/structures/RightPanel-test.tsx
index 9653fa6cd12..17e952e7fbd 100644
--- a/test/components/structures/RightPanel-test.tsx
+++ b/test/components/structures/RightPanel-test.tsx
@@ -70,9 +70,8 @@ describe("RightPanel", () => {
         // This is private but it's the only way to selectively enable stores
         await WidgetLayoutStore.instance.onReady();

-        // @ts-ignore
         // Make sure we start with a clean store
-        RightPanelStore.internalInstance = null;
+        RightPanelStore.instance.reset();
         RightPanelStore.instance.useUnitTestClient(cli);
         // @ts-ignore
         await RightPanelStore.instance.onReady();
```

### BUT HOW!? 
(not-so-TLDR)

The way the stores are structured means they are instantiated on import, generating an instance reference which then gets bound to the `window` object, just as a true singleton should. By resetting the `internalInstance` by hand then calling `useUnitTestClient()` we end up generating a new instance which detects the client on its own and uses that reference (because `MatrixClientPeg.get()` is mocked). Because it detects an instance, and because the singletons are now confused (there's two `RightPanelStore`s floating around: one instantiated and one attached to the `window`), `this` within `onReady` actually has a `ReadyWatchingStore` prototype, hence the undefined function call. 

The confusing part was trying to figure out why this didn't affect the `WidgetLayoutStore` given it's able to figure out which of the `onReady` functions to call, and with the correct context. It also has a `window`-stored reference created at the time of import, however its `internalInstance` wasn't being reset - if you add in a similar `internalInstance = null` for the `WidgetLayoutStore` then you'll mimic the bug the `RightPanelStore` was having.

To work around the issue, we just create a `reset()` function which obliterates state. For safety and to encourage developers to set sane defaults, we call that reset function in the constructor so that if we forget a class member during a refactoring then it'll have a high chance of breaking something (either at runtime or in tests) - eg if we forgot to reset `byRoom` then we might find conflicts between tests and therefore some angry error messages.

Comment (robintown) [reply src/stores/right-panel/RightPanelStore.ts:58]: Hmm, okay, thanks for investigating (and reminding me that arrow functions do not actually provide the desired override semantics). I still really can't follow this bit of the logic:

> Because it detects an instance, and because the singletons are now confused (there's two RightPanelStores floating around: one instantiated and one attached to the window), this within onReady actually has a ReadyWatchingStore prototype

i.e. how having two instances of a class would possibly impact the prototypes of those instances, even if they're semantically meant to be singletons. However I definitely see how your suggestion would be overall a more sane thing to do.

**Review by turt2live — APPROVED** [review]

File: src/stores/right-panel/RightPanelStore.ts (hunk 2 above)
Comment (turt2live) [reply src/stores/right-panel/RightPanelStore.ts]: that much I haven't figured out, but I think it's effectively call order and mocks getting confused, somehow.
